File: 1-centralized-logging/1-java/src/main/java/com/starci/logging/orders/OrdersService.java
package com.starci.logging.orders;

import java.util.Map;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/** Orders service — generates a trace and writes demo logs. */
@Service
public class OrdersService {

    private static final Logger log = LoggerFactory.getLogger(OrdersService.class);

    /** Simulate a successful order; log INFO with a trace for Loki lookup. */
    public Map<String, String> accept() {
        String trace = "trace-" + System.currentTimeMillis();
        // HOSTNAME in containers helps distinguish replicas in demos.
        String host = Optional.ofNullable(System.getenv("HOSTNAME")).orElse("local");
        log.info("[{}] Order accepted trace={}", host, trace);
        return Map.of("status", "accepted", "trace", trace);
    }

    /** Simulate an order failure; log ERROR then throw a real exception for Sentry. */
    public Map<String, String> fail() {
        String host = Optional.ofNullable(System.getenv("HOSTNAME")).orElse("local");
        log.error("[{}] Simulated order failure for Loki + Sentry demo", host);
        // Throw a real exception — the Sentry resolver catches this and reports to Sentry.
        throw new IllegalStateException("Simulated order processing failure");
    }
}

File: 1-centralized-logging/0-typescript/src/orders/orders.service.ts
/**
 * Orders service — generates a trace and writes demo logs.
 */
import { Injectable, Logger } from "@nestjs/common"

@Injectable()
export class OrdersService {
    private readonly logger = new Logger(OrdersService.name)

    /**
     * Simulate a successful order; log INFO with a trace for Loki lookup.
     */
    accept(): { status: string; trace: string } {
        const trace = `trace-${Date.now()}`
        // HOSTNAME in containers helps distinguish replicas in demos.
        this.logger.log(`[${process.env.HOSTNAME ?? "local"}] Order accepted trace=${trace}`)
        return { status: "accepted", trace }
    }

    /**
     * Simulate an order failure; log ERROR then throw a real Error.
     * SentryGlobalFilter catches the exception and reports it to Sentry.
     */
    fail(): never {
        this.logger.error(`[${process.env.HOSTNAME ?? "local"}] Simulated order failure for Loki + Sentry demo`)
        // Throw a real Error — SentryGlobalFilter catches this and reports to the Sentry dashboard.
        throw new Error("Simulated order processing failure")
    }
}
